nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/main/java/org/apache/nifi/json/JsonTreeRowRecordReader.java

```diff
@@ -44,30 +44,51 @@
 import java.util.function.Supplier;
 
 public class JsonTreeRowRecordReader extends AbstractJsonRowRecordReader {
-    private final RecordSchema schema;
 
+    private final RecordSchema schema;
 
     public JsonTreeRowRecordReader(final InputStream in, final ComponentLog logger, final RecordSchema schema,
-        final String dateFormat, final String timeFormat, final String timestampFormat) throws IOException, MalformedRecordException {
-        super(in, logger, dateFormat, timeFormat, timestampFormat);
-        this.schema = schema;
+                                   final String dateFormat, final String timeFormat, final String timestampFormat) throws IOException, MalformedRecordException {
+        this(in, logger, schema, dateFormat, timeFormat, timestampFormat, null, null, null);
     }
 
     public JsonTreeRowRecordReader(final InputStream in, final ComponentLog logger, final RecordSchema schema,
                                    final String dateFormat, final String timeFormat, final String timestampFormat,
-                                   final StartingFieldStrategy strategy, final String startingFieldName) throws IOException, MalformedRecordException {
-        super(in, logger, dateFormat, timeFormat, timestampFormat, strategy, startingFieldName);
-        this.schema = schema;
+                                   final StartingFieldStrategy startingFieldStrategy, final String startingFieldName,
+                                   final SchemaApplicationStrategy schemaApplicationStrategy)
+            throws IOException, MalformedRecordException {
+
+        super(in, logger, dateFormat, timeFormat, timestampFormat, startingFieldStrategy, startingFieldName);
+        if (startingFieldStrategy == StartingFieldStrategy.NESTED_FIELD && schemaApplicationStrategy == SchemaApplicationStrategy.WHOLE_JSON) {
+            this.schema = getSelectedSchema(schema, startingFieldName);
+        } else {
+            this.schema = schema;
+        }
+    }
+
+    private RecordSchema getSelectedSchema(final RecordSchema schema, final String startingFieldName) {
```

Comment:
This implementation can find only 1-level deep nested schemas, no deeper than that.

Apart from adjusting the production code, I propose the following test as well:
```
src/test/resources/json/single-element-deep-nested.json:
{
	"rootInt": 100,
	"rootString": "root_string",
	"nestedLevel1Record": {
		"nestedLevel1Int": 110,
		"nestedLevel1String": "root.level1:string",
		"nestedLevel2Record": {
			"nestedLevel2Int": 111,
			"nestedLevel2String": "root.level1.level2:string"
		}
	}
}

In org.apache.nifi.json.TestJsonTreeRowRecordReader:
    @Test
    void testStartFromDeepNestedObject() throws IOException, MalformedRecordException {
        String jsonPath = "src/test/resources/json/single-element-deep-nested.json";

        RecordSchema recordSchema = new SimpleRecordSchema(Arrays.asList(
                new RecordField("rootInt", RecordFieldType.INT.getDataType()),
                new RecordField("rootString", RecordFieldType.STRING.getDataType()),
                new RecordField("nestedLevel1Record", RecordFieldType.RECORD.getRecordDataType(
                        new SimpleRecordSchema(Arrays.asList(
                                new RecordField("nestedLevel1Int", RecordFieldType.INT.getDataType()),
                                new RecordField("nestedLevel1String", RecordFieldType.STRING.getDataType()),
                                new RecordField("nestedLevel2Record", RecordFieldType.RECORD.getRecordDataType(
                                        new SimpleRecordSchema(Arrays.asList(
                                                new RecordField("nestedLevel2Int", RecordFieldType.INT.getDataType()),
                                                new RecordField("nestedLevel2String", RecordFieldType.STRING.getDataType())
                                        ))
                                ))
                        ))
                ))
        ));

        SimpleRecordSchema expectedRecordSchema = new SimpleRecordSchema(Arrays.asList(
                new RecordField("nestedLevel2Int", RecordFieldType.INT.getDataType()),
                new RecordField("nestedLevel2String", RecordFieldType.STRING.getDataType())
        ));

        List<Object> expected = Collections.singletonList(
                new MapRecord(expectedRecordSchema, new HashMap<String, Object>() {{
                    put("nestedLevel2Int", 111);
                    put("nestedLevel2String", "root.level1.level2:string");
                }})
        );

        testReadRecords(jsonPath, recordSchema, expected, StartingFieldStrategy.NESTED_FIELD,
                "nestedLevel2Record", SchemaApplicationStrategy.WHOLE_JSON);
    }
```

Reply:
Indeed, I will change it to work with deeper nested fields as well. Thanks for the test case too!